### DISCUSSION TITLE
Question about CRLF keep-alive

### DISCUSSION THREAD

**Discussion by ww95211** [resiprocate/resiprocate#246] [Q&A]
Hi @sgodin , I have moved the question to  here.  Here are some pieces in GDB.
```
(gdb) pvector ((resip::SipMessage*)0xf0e33328).mBufferList 0
elem[0]: $1 = 0xcbf78e88 "\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n"
Vector size = 34940
Vector capacity = 65536
Element type = std::_Vector_base<char*, std::allocator<char*> >::pointer
```
We can see there is 34940 elements in mBufferList.

```
(gdb) pvector ((resip::SipMessage*)0xf0e33328).mBufferList 0 200
elem[0]: $2 = 0xcbf78e88 "\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n"
elem[1]: $3 = 0xcbf79698 "\r\n"
elem[2]: $4 = 0xcbf79ea8 "\r\n"
elem[3]: $5 = 0xcbf7c6f8 "\r\n"
elem[4]: $6 = 0xcbf7cf08 "\r\n"
elem[5]: $7 = 0xcbf949f8 "\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n"
elem[6]: $8 = 0xcbf95208 "\r\n"
elem[7]: $9 = 0xcbfaed38 "\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n"
elem[8]: $10 = 0xcbfaf548 "\r\n"
elem[9]: $11 = 0xcbfca8a8 "\r\n"
elem[10]: $12 = 0xcbfcb0b8 "\r\n"
elem[11]: $13 = 0xcbfe6c28 "\r\n"
elem[12]: $14 = 0xcbfe7438 "\r\n"
elem[13]: $15 = 0xcbc08118 "\r\n"
elem[14]: $16 = 0xcbc08928 "\r\n"
elem[15]: $17 = 0xcbc23478 "\r\n"
elem[16]: $18 = 0xcbc23c88 "\r\n"
elem[17]: $19 = 0xcbc3e7d8 "\r\n"
elem[18]: $20 = 0xcbc3efe8 "\r\n"
elem[19]: $21 = 0xcbc59b38 "\r\n"
elem[20]: $22 = 0xcbc5a348 "\r\n"
elem[21]: $23 = 0xcbc74e98 "\r\n"
elem[22]: $24 = 0xcbc756a8 "\r\n"
elem[23]: $25 = 0xcbc901f8 "\r\n"
elem[24]: $26 = 0xcbc90a08 "\r\n"
elem[25]: $27 = 0xcbcab558 "\r\n"
elem[26]: $28 = 0xcbcabd68 "\r\n"
elem[27]: $29 = 0xcbcc68b8 "\r\n"
elem[28]: $30 = 0xcbcc70c8 "\r\n"
elem[29]: $31 = 0xcbce1c18 "\r\n"
elem[30]: $32 = 0xcbce2428 "\r\n"
elem[31]: $33 = 0xcbcfcf78 "\r\n"
elem[32]: $34 = 0xcbcfd788 "\r\n"
elem[33]: $35 = 0xcbd1b3a8 "\r\n"
elem[34]: $36 = 0xcbd1bbb8 "\r\n"
elem[35]: $37 = 0xcbd36708 "\r\n"
elem[36]: $38 = 0xcbd36f18 "\r\n"
elem[37]: $39 = 0xcbd4ea08 "\r\n"
elem[38]: $40 = 0xcbd4f218 "\r\n"
elem[39]: $41 = 0xcbd69d68 "\r\n"
```
There are lot of CRLFs in SipMessage.mBufferList,  I don't know why a new SipMessage is created to handle in this case. I wrote a tool to send CRLFs in short time with high rate, the similar behavior is observed also. I think  it may be caused by TCP subpackaging. In ConnectionBase::preparseNewBytes, if one connection receives CR and then LF, it is that to say CR and LF is not received at the same, it could new a SipMessage and mConnState is changed from NewMessage to ReadingHeaders. So SipMessage.mBufferList size will be increased to a huge value.

**Comment by sgodin** [comment]
Hmm - seems like the TCP framing somehow got off.  Not sure offhand what transpired to get into this situation.  

There was recently a bug fix checked in on March 24, 2022 related to TCP transports and CRLF's, I wonder if you are hitting it:
https://github.com/resiprocate/resiprocate/commit/7bc80ff508855a2304154d0ee00b6d1927574b08

Are you using the latest code?

Scott

**Comment by ww95211** [comment]
Thanks @sgodin , I am not using the latest code. With the latest code, the problem is fixed. But if I send a CR and then LF, the issue still exists. after I add a protection for SipMessage:mBufferList's size, this problem is fixed. Thanks again! @sgodin

**Reply by pengweichu** [reply]
> Thanks @sgodin , I am not using the latest code. With the latest code, the problem is fixed. But if I send a CR and then LF, the issue still exists. after I add a protection for SipMessage:mBufferList's size, this problem is fixed. Thanks again! @sgodin

Could you share your protection? Thanks.
